db/db_impl/db_impl_merge.cc

```diff
@@ -0,0 +1,361 @@
+        if (!cfd->mem()->IsEmpty()) {
+          WriteContext write_context;
+          assert(log_empty_);
+          s = SwitchMemtable(cfd, &write_context);
+          if (!s.ok()) {
+            return s;
+          }
+        }
+        assert(cfd->mem()->IsEmpty());
+
+        // [B] Bump log number for active memtable. Even though it's not
+        // shared, it must still be larger than other shared immutable
+        // memtables.
+        cfd->mem()->SetNextLogNumber(max_log_number);
+        cfd->imm()->ExportMemtables(&mems);
+      }
+
+      // Acquire super version.
+      cf_db_super_versions[cf_i][db_i] = cfd->GetSuperVersion()->Ref();
+    }
+    for (auto mem : mems) {
+      assert(mem != nullptr);
+      mem->Ref();
+      // [B] Bump log number for shared memtables.
+      mem->SetNextLogNumber(max_log_number);
+      this_cfd->imm()->Add(mem, &to_delete);
+    }
+    this_cfd->mem()->SetNextLogNumber(max_log_number);
+  }
+  for (size_t i = 0; i < all_db_impls.size(); i++) {
+    auto* db = all_db_impls[i];
+    bool check_log_number = (i == 0 || merge_options.allow_source_write) &&
+                            merge_options.merge_memtable;
+    if (check_log_number && max_log_number != db->logfile_number_) {
+      assert(max_log_number > db->logfile_number_);
+      // [B] Create a new WAL so that future memtable will use the correct log
+      // number as well.
+      log::Writer* new_log = nullptr;
+      s = db->CreateWAL(max_log_number, 0 /*recycle_log_number*/,
+                        0 /*preallocate_block_size*/, &new_log);
+      if (!s.ok()) {
+        return s;
+      }
+      db->logfile_number_ = max_log_number;
+      assert(new_log != nullptr);
+      db->logs_.emplace_back(max_log_number, new_log);
+      auto current = db->versions_->current_next_file_number();
+      if (current <= max_log_number) {
+        db->versions_->FetchAddFileNumber(max_log_number - current + 1);
+      }
+    }
+  }
+
+  // Unblock writes.
+  write_blockers.clear();
+
+  TEST_SYNC_POINT("DBImpl::MergeDisjointInstances:AfterMergeMemtable:1");
+
+  // # Merge table files
+  assert(s.ok());
+  autovector<VersionEdit> cf_edits;
+  cf_edits.resize(num_cfs);
+  for (size_t cf_i = 0; cf_i < num_cfs; cf_i++) {
+    auto* this_cfd = this_cfds[cf_i];
+    auto& edit = cf_edits[cf_i];
+    edit.SetColumnFamily(this_cfd->GetID());
+    for (size_t db_i = 0; db_i < db_impls.size(); db_i++) {
+      auto* super_version = cf_db_super_versions[cf_i][db_i];
+      if (super_version == nullptr) {
+        continue;
+      }
+      VersionStorageInfo& vsi = *super_version->current->storage_info();
+      auto& cf_paths = super_version->cfd->ioptions()->cf_paths;
+      auto SourcePath = [&](size_t path_id) {
+        // Matching `TableFileName()`.
+        if (path_id >= cf_paths.size()) {
+          assert(false);
+          return cf_paths.back().path;
+        } else {
+          return cf_paths[path_id].path;
+        }
```

Comment:
Why not just:
```
assert(path_id < cf_paths.size());
return cf_paths[path_id].path;
```?

Reply:
This code is copies in exact form as the original ("Matching `TableFileName()`").